### PR TITLE
TFP: champion config re-verification post cp_panel fix (seed=0/42)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #3209 fixed the cp_panel_prior_index() bug that caused the TFP champion config to produce Inf pressure values for most seeds. The bug was that cp_panel_prior_index() returned a non-None index for tandemfoil_paper even though build_tandem_paper_bundle() never populates cp_panel features — causing sinh() to be applied to a Fourier feature column value instead of a proper cp_panel value.

With the fix now merged into radford, we can re-verify the champion config. The goal is to confirm that the TFP baseline of val_primary/field_mse = 0.002383 (PR #3025) is reproducible under the current codebase, and potentially improve on it.

**This is the HIGHEST PRIORITY TFP experiment.** The entire TFP benchmark has been blocked since #3205 confirmed the code regression.

### Trial 1 — Champion config reproduction (seed=0)

Run the exact PR #3025 champion config with seed=0:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --epochs 999 \
  --ema-decay 0.999 \
  --seed 0 \
  --wandb_group tfp-champion-rerun-post-bugfix
```

**Note:** Do NOT add --asinh-pressure, --residual-prediction, --enable-cp-panel, --enable-pressure-prior-addition. The champion config (#3025) used the simpler config without these flags. The cp_panel guard in #3209 makes these safe, but the champion was trained WITHOUT them.

### Trial 2 — Seed=42 (reproducibility check)

Same command but with `--seed 42` to confirm cross-seed stability:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --epochs 999 \
  --ema-decay 0.999 \
  --seed 42 \
  --wandb_group tfp-champion-rerun-post-bugfix
```

### What to report

1. val_primary/field_mse and test_primary/field_mse for both seeds
2. Whether either seed achieves < 0.002383 (the old champion)
3. Whether both seeds produce finite metrics throughout training (no Inf/NaN)
4. The training curve — does it converge smoothly or show instabilities?

## Baseline

TFP baseline (PR #3025, pre-code-regression):

| Metric | Value |
|--------|-------|
| val_primary/field_mse | **0.002383** |

This was achieved with the above config at seed=0 before the cp_panel bug was introduced. It has been unreproducible since.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT sanji:

## Results

### Trial 1 — Seed=0 (W&B run: `rglchxrv`)

| Metric | Value | vs Baseline (0.002383) |
|--------|-------|----------------------|
| val_primary/field_mse | **0.00474** (ep287) | +99% (worse) |
| test_primary/field_mse | **0.00437** (best-ckpt) | — |
| test/surface_mse | 0.00348 | — |
| test/volume_mse | 0.00438 | — |
| test/surface_mse_p | 0.00523 | — |

- **Epochs trained:** 538 (360-min timeout)
- **Best epoch:** 287
- **Convergence:** Smooth descent to ep287 best, then gradual degradation to val ~0.035 by ep538 (T_max=10 cycling instability). No catastrophic divergence.
- **Finite metrics:** All val metrics finite throughout training — **no Inf/NaN**.
- **EMA:** Correctly preserved best checkpoint at ep287.

### Trial 2 — Seed=42 (W&B run: `7eox2joe`)

| Metric | Value | vs Baseline (0.002383) |
|--------|-------|----------------------|
| val_primary/field_mse | **0.00547** (ep205) | +130% (worse) |
| test_primary/field_mse | **0.00522** (best-ckpt) | — |
| test/surface_mse | 0.00349 | — |
| test/volume_mse | 0.00525 | — |
| test/surface_mse_p | 0.00537 | — |

- **Epochs trained:** 548 (360-min timeout)
- **Best epoch:** 205
- **Convergence:** Smooth descent to ep205, then gradual degradation ep205–498. **Catastrophic divergence** at ep499–500: loss 0.34→1.2→17+, grad norms → 10^16 → Infinity. val_primary/field_mse exploded to 13.4 by final epoch.
- **Finite metrics during training phase:** Yes, all val metrics finite through ep498. Inf grad norms from ep519 onwards.
- **EMA:** Correctly preserved best checkpoint at ep205.

### Summary

| Question | Answer |
|----------|--------|
| Does either seed achieve < 0.002383? | **No.** Seed=0 got 0.00474, seed=42 got 0.00547 — both ~2x worse. |
| Are both seeds finite throughout? | Seed=0 yes. Seed=42 finite through ep498, then catastrophic divergence. |
| Training curve stable? | Both show T_max=10 cycling instability after best epoch. Seed=42 diverges catastrophically; seed=0 degrades gradually. |
| Cross-seed stability? | Moderate — both in the 0.0047–0.0055 range, but seed=42 finds best 82 epochs earlier and diverges more violently. |

### Important observation: config mismatch with BASELINE.md

The PR instructions say to NOT use `--asinh-pressure`, `--residual-prediction`, `--enable-cp-panel`, `--enable-pressure-prior-addition`. However, the **BASELINE.md reproduce command** for the TFP champion (PR #3025) **includes all of these flags plus `--enable-te-coord-frame` and `--enable-cp-panel-tandem-only`**.

The BASELINE.md config reads: _"Lion lr=1.25e-4, T_max=10, gc=0.5, WD=1e-2, EMA=0.999, 3L/192d, **Fourier+physics**"_ — and the reproduce command explicitly includes the physics features.

This likely explains the 2x gap: the original champion was trained WITH physics features, but these runs used the simpler config WITHOUT them. I'd recommend re-running with the full BASELINE.md reproduce command to properly verify the post-bugfix champion.

### Commands used

```bash
# Seed=0 (GPU 0)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --epochs 999 --ema-decay 0.999 --seed 0 \
  --agent sanji --wandb-name "sanji/tfp-champion-rerun-seed0" \
  --wandb-group tfp-champion-rerun-post-bugfix

# Seed=42 (GPU 1) — same but --seed 42
```

### Suggested follow-ups

1. **Re-run with full BASELINE.md physics flags** (`--enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition`) to properly test whether the 0.002383 baseline is reproducible post-bugfix.
2. **Test gc=0.3** to see if softer gradient clipping helps stability at longer training horizons (both seeds diverged well before timeout).
3. **T_max=20 or T_max=30** for better late-training stability — T_max=10 cycling is the divergence trigger in both seeds.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3266 because ADVISOR: Closing — val=0.00474 is 2x worse than baseline 0.002344. Sanji reassigned..
